File: code/java/ch09/practice/Challenge03RotatedSearchII.java
package ch09.practice;

import java.util.*;

/**
 * Challenge 03: Search in Rotated Sorted Array II (with duplicates)
 * ==============================
 * Chapter 9: Finding Needles — The Power of Searching
 *
 * PROBLEM: Given a rotated sorted array that MAY CONTAIN DUPLICATES
 *          and a target, return true if the target exists, false otherwise.
 *
 * ALGORITHM: Modified binary search for rotated arrays. When duplicates
 *            make it impossible to determine which half is sorted
 *            (arr[lo]==arr[mid]==arr[hi]), shrink from both ends.
 *
 * EXAMPLES:
 *   solve(new int[]{2,5,6,0,0,1,2}, 0)  = true
 *   solve(new int[]{2,5,6,0,0,1,2}, 3)  = false
 *   solve(new int[]{1,0,1,1,1}, 0)      = true
 *
 * CONSTRAINTS: 1 <= arr.length <= 10^5, may contain duplicates
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Challenge03RotatedSearchII {
    public static boolean solve(int[] arr, int target) {
        // TODO: Replace this with your solution
        return false;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine().trim();
        int[] arr;
        if (line.isEmpty()) {
            arr = new int[0];
        } else {
            String[] parts = line.split("\\s+");
            arr = new int[parts.length];
            for (int i = 0; i < parts.length; i++) arr[i] = Integer.parseInt(parts[i]);
        }
        int target = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(arr, target));
        sc.close();
    }
}

File: code/cpp/ch09/practice/challenge_03_rotated_search_ii.cpp
/*
 * Challenge 3: Search in Rotated Sorted Array II (with duplicates)
 * ==================================================================
 * Chapter 9: Finding Needles -- The Power of Searching
 *
 * PROBLEM:
 *   Given a rotated sorted array that MAY CONTAIN DUPLICATES and a
 *   target, return true if target exists in the array, false otherwise.
 *
 * EXAMPLES:
 *   solve({2,5,6,0,0,1,2}, 0) -> true
 *   solve({2,5,6,0,0,1,2}, 3) -> false
 *   solve({1,0,1,1,1}, 0)     -> true
 *   solve({1,1,1,1,1}, 2)     -> false
 *   solve({1}, 1)             -> true
 *
 * CONSTRAINTS:
 *   1 <= arr.size() <= 10^5
 *   Array may contain duplicates
 *
 * NOTE:
 *   Worst case is O(n) due to duplicates, but average is O(log n).
 *   When arr[lo] == arr[mid] == arr[hi], we can't determine which
 *   half is sorted, so we shrink both ends.
 *
 * INSTRUCTIONS:
 *   Replace the body with your solution.
 */

#include <iostream>
#include <vector>
using namespace std;

bool solve(vector<int> arr, int target) {
    // TODO: Replace this with your solution
    return false;
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    int target;
    cin >> target;
    cout << (solve(arr, target) ? "true" : "false") << endl;
    return 0;
}
